app/controllers/voucher_adjustments_controller.rb

```diff
@@ -59,6 +78,35 @@ def add_voucher
     true
   end
 
+  def load_voucher
+    voucher = Voucher.not_vine.find_by(code: voucher_params[:voucher_code],
+                                       enterprise: @order.distributor)
+    return voucher unless voucher.nil?
+
+    vine_voucher
+  end
+
+  def vine_voucher
+    vine_voucher_validator = VineVoucherValidatorService.new(
+      voucher_code: voucher_params[:voucher_code], enterprise: @order.distributor
+    )
+    voucher = vine_voucher_validator.validate
+
+    if vine_voucher_validator.errors.present?
+      return nil if ignored_vine_api_error(vine_voucher_validator)
+
+      @order.errors.add(:voucher_code, I18n.t('checkout.errors.add_voucher_error'))
+      return nil
+    end
```

Comment:
I had to parse all the files in my head to figure out what was happening. From what I can understand, a simpler way to say this is:

```suggestion
    # if certain errors are present, add an extra message to the order
    if vine_voucher_validator.errors.keys.intersect? :invalid_voucher, :vine_api
      @order.errors.add(:voucher_code, I18n.t('checkout.errors.add_voucher_error'))
    end
```

I think the `voucher_not_found` error will still be applied in this case, which seems redundant. Maybe we need to be applying the error message a bit further up the call stack.

Reply:
I would work, but then you loose some context. I want to ignore some error, not allow only `:invalid_voucher` and :vine_api . What if we add a new error to the validator that we don't want to ignore ? then we need to update this code. No need to do that with my solution. 
I do concede it looks a bit ugly, and maybe we need to handle the errors somewhere else.